OSDOCS-3903: known issue BZ2100323, update bug fix (New Jira Ticket) (#48796)
openshift/openshift-docs: release_notes/ocp-4-11-release-notes.adoc

* Previously, pod failures were artificially extending the validity period of certificates causing them to incorrectly rotate. With this update, the certificate validity period is correctly determined and the certificates are correctly rotated. (link:https://bugzilla.redhat.com/show_bug.cgi?id=2020484[*BZ#2020484*])

> * In {product-title} {product-version} the default cluster-wide pod security admission policy is set to `privileged` for all namespaces and the default warning level is set to `restricted`. Before this update, Operator Lifecycle Manager displayed pod security admission warnings in the `operator-marketplace` namespace. With this fix, reducing the warning level to `baseline` resolves the issue. (link:https://bugzilla.redhat.com/show_bug.cgi?id=2088541[*BZ#2088541*])

Reviewer: > Before this update,

How to understand it? What's this `update` means? The PSA? If yes, should be `after this update`.
> warnings in the `operator-marketplace` namespace.

And, should be `openshift-marketplace`, not the `operator-marketplace`.

Author: I changed the namespace to `openshift-marketplace`. 

Before this update refers to BZ2088541. This is just a bug fix note for the doc text supplied in the BZ. Though I have cleaned it up to conform with our style guide. I noticed that the doc text incorrectly said that baseline was the default, so I was fixing it here.

Author: "Before this update" is referring to the time before the bug was fixed. "With this fix" is referring to the time after the BZ was fixed.

> * In {product-title} {product-version} the default cluster-wide pod security admission policy is set to `privileged` for all namespaces and the default warning level is set to `restricted`. Before this update, Operator Lifecycle Manager displayed pod security admission warnings in the `operator-marketplace` namespace. With this fix, reducing the warning level to `baseline` resolves the issue. (link:https://bugzilla.redhat.com/show_bug.cgi?id=2088541[*BZ#2088541*])

Reviewer 2: ```suggestion
* In {product-title} {product-version} the default cluster-wide pod security admission policy is set to `privileged` for all namespaces which are not part of the OpenShift system and the default warning level is set to `restricted`. Before this update, Operator Lifecycle Manager displayed pod security admission warnings in the `operator-marketplace` namespace. With this fix, reducing the warning level to `baseline` resolves the issue. (link:https://bugzilla.redhat.com/show_bug.cgi?id=2088541[*BZ#2088541*])
```

@michaelryanpeter  all namespaces prefixed with openshift and + a few of them are restricted but I do not think we need to describe them here. More info: 
[OCP 4.11 Documentation](https://github.com/openshift/openshift-docs/pull/48409)
[OCP 4.11 Release Notes](https://github.com/openshift/openshift-docs/pull/48402/files)

Author: This is just a summary and rewording of the doc text in the BZ. This is not meant to do anything other than provide a more customer friendly summary of the content of the doc text in the BZ.

Reviewer 2: @michaelryanpeter the info is not accurate and I think we should link here the release notes that explain what was changed in the OCP 4.11 and why the warning is raised now.

> * In {product-title} {product-version} the default cluster-wide pod security admission policy is set to `privileged` for all namespaces and the default warning level is set to `restricted`. Before this update, Operator Lifecycle Manager displayed pod security admission warnings in the `operator-marketplace` namespace. With this fix, reducing the warning level to `baseline` resolves the issue. (link:https://bugzilla.redhat.com/show_bug.cgi?id=2088541[*BZ#2088541*])

Reviewer 2: ```suggestion
* In {product-title} {product-version} the default cluster-wide pod security admission policy is set to `privileged` for all namespaces and the default warning level is set to `restricted`. Before this update, Operator Lifecycle Manager displayed pod security admission warnings in the `operator-marketplace` namespace. Therefore, the warning level on this specific namespace was changed by default to `baseline`. (link:https://bugzilla.redhat.com/show_bug.cgi?id=2088541[*BZ#2088541*])
```

Why the warning is an issue?
@perdasilva

* The pipeline metrics API does not support the required `pipelinerun/taskrun` histogram values from {rh-openstack} 1.6 and beyond. Consequently, the *Metrics* tab in the *Pipeline* -> *Details* page is removed instead of displaying incorrect values. There is currently no workaround for this issue. (link: https://bugzilla.redhat.com/show_bug.cgi?id=2074767[*BZ#2074767*])

> * In {product-title} {product-version}, pod security admission runs globally with privileged enforcement and restricted audit logging and API warnings. Versions of the `opm` binary released earlier than {product-title} {product-version}, do not work with the restricted profile setting. As a result, an error displays if you install a catalog source that was built with an `opm` binary released earlier than {product-title} {product-version}.

Author: The note that "pod security admission runs globally with privileged enforcement and restricted audit logging and API warnings" comes from the PSA release note above.

Reviewer: But, why do we still mention the `opm` here? As I said on Slack, no matter the Catalog Source created by the new or old `opm`, both of them will encounter the PSA issue, because:

- registry-server (catalog source pod’s container, but no matter the SQL-based or Filed based, both of them have this issue)

- unpack job (nothing with the CatalogSource, but payload)

Reviewer 2: @michaelryanpeter @jianzhangbjz:
The error here occurs because:

Previous images ( no matter the format ) built with OPM < 1.23 (please confirm the version) does not define the UserID (i.e. USER 65532:65532  ) in the Dockerfile. Therefore, they cannot work with the security context RunAsNonRoot set as true which is a requirement to comply with the Pod Security standards. 

Could we improve this one and add a better description for it? 
@anik120 ^

> * In {product-title} {product-version}, pod security admission runs globally with privileged enforcement and restricted audit logging and API warnings. Versions of the `opm` binary released earlier than {product-title} {product-version}, do not work with the restricted profile setting. As a result, an error displays if you install a catalog source that was built with an `opm` binary released earlier than {product-title} {product-version}.

Reviewer 2: ```suggestion
* In {product-title} {product-version}, Pod Security Admission runs globally with privileged enforcement and restricted audit logging and API warnings. Versions of the `opm` binary released earlier than {product-title} {product-version}, do not defines the UserID on the image (i.e. in the Dockerfile we have not `USER 11211:11211 `) and in this will fail when qualified to the restricted-v2 SCC policy where the RunAsNonRoot will be set with true value as part of its strategy. 

In this way, be aware that index images built with an `opm` binary released earlier than {product-title} {product-version} before this issue can fail with the following error if applied on namespaces enforced as restricted and with the pod security admission label sychronization is disabled.
```

Reviewer 2: @perdasilva @anik120 ^ WDYT?

> * In {product-title} {product-version}, pod security admission runs globally with privileged enforcement and restricted audit logging and API warnings. Versions of the `opm` binary released earlier than {product-title} {product-version}, do not work with the restricted profile setting. As a result, an error displays if you install a catalog source that was built with an `opm` binary released earlier than {product-title} {product-version}.

Reviewer 2: /hold

> * In {product-title} {product-version}, pod security admission runs globally with privileged enforcement and restricted audit logging and API warnings. Versions of the `opm` binary released earlier than {product-title} {product-version}, do not work with the restricted profile setting. As a result, an error displays if you install a catalog source that was built with an `opm` binary released earlier than {product-title} {product-version}.

Reviewer 2: > * In {product-title} {product-version}, pod security admission runs globally with privileged enforcement and restricted audit logging and API warnings. 

@michaelryanpeter @anik120 @jianzhangbjz   ^ the above text has no relation to the issue ^. Btw, all are still as "Previlged" and that will only change on 4.12. From 4.12, Openshift system namespaces (whch are these ones [here](https://github.com/openshift/openshift-docs/pull/48409/files#diff-6b079749a009ff430faa75c3cdece0921ba0e269d1dcd32d6ec0939c7ff99059R9-R23)) will be enforced to restricted. 

Following my suggestion

> * `opm` binary released earlier than {product-title} {product-version}, do not work with the restricted profile setting. As a result, an error displays if you install a catalog source that was built with an `opm` binary released earlier than {product-title} {product-version} will be raised when the namespace is enforced to be restricted and the pod security admission label synchronization is disabled. Note that Openshift system namespaces do not allow you to enable the pod security admission label synchronization.  (We need a link here to the SCC docs, see https://github.com/openshift/openshift-docs/pull/48409/files#diff-6b079749a009ff430faa75c3cdece0921ba0e269d1dcd32d6ec0939c7ff99059R5-R23) 

@anik120 did we not revert the changes? 
Were not the error faced here because before we did not set

> * In {product-title} {product-version}, pod security admission runs globally with privileged enforcement and restricted audit logging and API warnings. Versions of the `opm` binary released earlier than {product-title} {product-version}, do not work with the restricted profile setting. As a result, an error displays if you install a catalog source that was built with an `opm` binary released earlier than {product-title} {product-version}.

Reviewer 2: ```suggestion
* `opm` binary released earlier than {product-title} {product-version}, do not work with the restricted profile setting. As a result, an error displays if you install a catalog source that was built with an `opm` binary released earlier than {product-title} {product-version} will be raised when the namespace is enforced to be restricted and the pod security admission label synchronization is disabled. Note that Openshift system namespaces do not allow you to enable the pod security admission label synchronization.  (We need a link here to the SCC docs, see https://github.com/openshift/openshift-docs/pull/48409/files#diff-6b079749a009ff430faa75c3cdece0921ba0e269d1dcd32d6ec0939c7ff99059R5-R23) 
```

@michaelryanpeter @anik120 @jianzhangbjz

+
.Example error
[source,terminal]
----

> Error: open ./db-xxxx: permission denied

Reviewer: Sorry, I think this error is nothing with the PSA, it's related to bug: https://bugzilla.redhat.com/show_bug.cgi?id=2100323 or am I missing something? Thanks!

Author: I am trying to supply the known issue and work around for BZ2100323. The BZ is linked on line 2471.

----
+

> Workaround: Cluster administrators can allow the creation of catalog sources built with earlier `opm` versions by setting the pod security admission labels in the namespace metadata files. Apply the `baseline` label to the `warn`, `audit`, and `enforce` modes and set the `scc.podSecurityLabelSync` label to `false`, as shown in the following example:

Reviewer 2: ```suggestion
Workaround: Cluster administrators can allow the creation of catalog sources built with earlier `opm` versions by setting enabling the pod security admission label sychronization in a namespace, setting the value of the `security.openshift.io/scc.podSecurityLabelSync` label to `true`.

Run the following command:
+
[source,terminal]
----
$ oc label namespace <namespace> security.openshift.io/scc.podSecurityLabelSync=true
----

**WARNING:**  The above workaround does not work with the OCP system namespace. More info:  

[NOTE]
====
By default, user-created namespaces that have the prefix `openshift-` have pod security admission label synchronization disabled.
Namespaces that the installer creates have pod security admission label synchronization disabled permanently. These namespaces include:
* All namespaces that are prefixed with `openshift-`, except for `openshift-operators`
* `default`
* `kube-node-lease`
* `kube-system`
* `kube-public`
* `openshift`
====
```
@michaelryanpeter 
See: https://github.com/openshift/openshift-docs/pull/48409/files#diff-6b079749a009ff430faa75c3cdece0921ba0e269d1dcd32d6ec0939c7ff99059R10-R23

Reviewer 2: @perdasilva @anik120 ^

> Workaround: Cluster administrators can allow the creation of catalog sources built with earlier `opm` versions by setting the pod security admission labels in the namespace metadata files. Apply the `baseline` label to the `warn`, `audit`, and `enforce` modes and set the `scc.podSecurityLabelSync` label to `false`, as shown in the following example:

Reviewer 2: @michaelryanpeter a warning can not cause or fix any issue.
@anik120

> Workaround: Cluster administrators can allow the creation of catalog sources built with earlier `opm` versions by setting the pod security admission labels in the namespace metadata files. Apply the `baseline` label to the `warn`, `audit`, and `enforce` modes and set the `scc.podSecurityLabelSync` label to `false`, as shown in the following example:

Reviewer 2: @michaelryanpeter @anik120:

It is wrong:

a) All namespaces are sync by default with ``security.openshift.io/scc.podSecurityLabelSync` unless those that are considered Openshift system namespaces. Please, see the SCC doc: https://github.com/openshift/openshift-docs/pull/48409/files#diff-a9fd77a91dcedf5db05ac33bca1bcd89e2c8c2e4b444b6dbdf0f6c06cba1af53R19-R35
b) Therefore, we should NOT recommend users create CatalogSource or Operator in Openshift system namespaces
c) The namespace openshift-marketplace is ONLY used by OLM and OLM labels it.

So, we need to clarify it here:

a) If you want to create a new Catalogue Source it is not recommended use the Openshift system namespace
b) If you create a CatalogSource in a namespace prefixed with openshift and you face the error because you are using an image that was built with the old OPM tool then, ensure that you enable the PSA label synchronization. We should link  SCC doc here: https://github.com/openshift/openshift-docs/pull/48409/files#diff-6b079749a009ff430faa75c3cdece0921ba0e269d1dcd32d6ec0939c7ff99059R30-R44

@jianzhangbjz ^

Reviewer: >  If you want to create a new Catalogue Source it is not recommended use the Openshift system namespace

If the customer create a CatalogSource in their namespace, they will encounter bug https://bugzilla.redhat.com/show_bug.cgi?id=2088541, the workaround is add labels for their namespace:
```yaml
apiVersion: v1
kind: Namespace
metadata:
...
  labels:
    security.openshift.io/scc.podSecurityLabelSync: "false"
    openshift.io/cluster-monitoring: "true"
    pod-security.kubernetes.io/audit: baseline <1>
    pod-security.kubernetes.io/warn: baseline <2>
    pod-security.kubernetes.io/enforce: baseline <3>
  name: "<namespace_name>"
```

Reviewer 2: @jianzhangbjz the warning is NOT a bug. It was tracked as a bug with the expectation that all that is shipped/installed by default via OLM does not raise warnings. 

The workarround here is about the error `Error: open ./db-xxxx: permission denied` and not the warning.
Therefore, the workarround here is not right and we need to clarify that users should NOT create catalog sources in Openshuft system namespaces and if they create a new namespace prefixed with openshift they also should enable the PSA label syncer instead. 

c/c @michaelryanpeter @anik120

Reviewer: > The workarround here is about the error Error: open ./db-xxxx: permission denied

Hi @camilamacedo86 , thanks! But, why do we need this workaround? Based on https://bugzilla.redhat.com/show_bug.cgi?id=2100323#c4, it had been fixed.

Reviewer: This index image `quay.io/olmqe/etcd-index:v1` was created by using the old `opm`, but I don't encounter this `Error: open ./db-xxxx: permission denied` error when installing it on the latest OCP 4.11 cluster. As follows,
```yaml
MacBook-Pro:~ jianzhang$ oc get clusterversion
NAME      VERSION                              AVAILABLE   PROGRESSING   SINCE   STATUS
version   4.11.0-0.nightly-2022-08-04-081314   True        False         3h30m   Cluster version is 4.11.0-0.nightly-2022-08-04-081314
MacBook-Pro:~ jianzhang$ cat cs-test.yaml 
apiVersion: operators.coreos.com/v1alpha1
kind: CatalogSource
metadata:
  name: test-operators
  namespace: openshift-marketplace
spec:
  displayName: Jian Operators
  image: quay.io/olmqe/etcd-index:v1
  priority: -100
  publisher: Jian
  sourceType: grpc
  updateStrategy:
    registryPoll:
      interval: 10m0s

MacBook-Pro:~ jianzhang$ oc get catalogsource
NAME                  DISPLAY                TYPE   PUBLISHER      AGE
certified-operators   Certified Operators    grpc   Red Hat        3h46m
community-operators   Community Operators    grpc   Red Hat        3h46m
qe-app-registry       Production Operators   grpc   OpenShift QE   3h24m
redhat-marketplace    Red Hat Marketplace    grpc   Red Hat        3h46m
redhat-operators      Red Hat Operators      grpc   Red Hat        3h46m
test-operators        Jian Operators         grpc   Jian           2m23s

MacBook-Pro:~ jianzhang$ oc get pods
NAME                                                              READY   STATUS      RESTARTS        AGE
1263538589d348af98ee1c1a3af1e63c7d8f8f7b9bf1a043d121dac5c9c92fk   0/1     Completed   0               3h21m
3ef707a2bab80ae871eaa876bd55216b08d6a4980ebb5b8ed5cd9c123d7rpqf   0/1     Completed   0               3h21m
certified-operators-wnwzw                                         1/1     Running     0               3h20m
community-operators-5tr9l                                         1/1     Running     0               122m
marketplace-operator-5b84d4d799-4scrc                             1/1     Running     1 (3h39m ago)   3h50m
qe-app-registry-kl8gq                                             1/1     Running     0               3h22m
redhat-marketplace-4bgs9                                          1/1     Running     0               3h43m
redhat-operators-mc46k                                            1/1     Running     0               3h43m
test-operators-dl4xf                                              1/1     Running     0               12s
MacBook-Pro:~ jianzhang$ oc rsh test-operators-dl4xf
/ # ps -elf|cat
PID   USER     TIME  COMMAND
    1 root      0:00 /bin/opm registry serve --database /database/index.db
   78 root      0:00 /bin/sh
   84 root      0:00 ps -elf
   85 root      0:00 cat
/ # exit

```

Reviewer: ```yaml
MacBook-Pro:~ jianzhang$ oc get ns openshift-marketplace -o yaml
apiVersion: v1
kind: Namespace
metadata:
  annotations:
    capability.openshift.io/name: marketplace
    include.release.openshift.io/ibm-cloud-managed: "true"
    include.release.openshift.io/self-managed-high-availability: "true"
    include.release.openshift.io/single-node-developer: "true"
    openshift.io/node-selector: ""
    openshift.io/sa.scc.mcs: s0:c17,c9
    openshift.io/sa.scc.supplemental-groups: 1000290000/10000
    openshift.io/sa.scc.uid-range: 1000290000/10000
    workload.openshift.io/allowed: management
  creationTimestamp: "2022-08-08T23:14:30Z"
  labels:
    kubernetes.io/metadata.name: openshift-marketplace
    olm.operatorgroup.uid/a93b3ad3-dc83-4801-908e-65b1de607ded: ""
    openshift.io/cluster-monitoring: "true"
    pod-security.kubernetes.io/audit: baseline
    pod-security.kubernetes.io/warn: baseline
  name: openshift-marketplace
  ownerReferences:
  - apiVersion: config.openshift.io/v1
    kind: ClusterVersion
    name: version
    uid: d9e8b196-691f-4e8e-92c8-c83338f46eec
  resourceVersion: "8735"
  uid: ecc126e9-1d9e-43bd-8c40-7cb81c566497
spec:
  finalizers:
  - kubernetes
status:
  phase: Active
```

Reviewer: I also install it in a `restricted` project, but still not encounter this `Error: open ./db-xxxx: permission denied`. So, I'm confused why we treat this `Error: open ./db-xxxx: permission denied` as a known issue, it had been fixed.
```yaml
MacBook-Pro:~ jianzhang$ oc get catalogsource -n test
NAME             DISPLAY          TYPE   PUBLISHER   AGE
test-operators   Jian Operators   grpc   Jian        14s
MacBook-Pro:~ jianzhang$ oc get pods -n test
NAME                   READY   STATUS    RESTARTS   AGE
test-operators-k9wsc   1/1     Running   0          21s
MacBook-Pro:~ jianzhang$ 
MacBook-Pro:~ jianzhang$ oc get ns test -o yaml
apiVersion: v1
kind: Namespace
metadata:
  annotations:
    openshift.io/description: ""
    openshift.io/display-name: ""
    openshift.io/sa.scc.mcs: s0:c27,c14
    openshift.io/sa.scc.supplemental-groups: 1000730000/10000
    openshift.io/sa.scc.uid-range: 1000730000/10000
  creationTimestamp: "2022-08-09T03:21:26Z"
  labels:
    kubernetes.io/metadata.name: test
    pod-security.kubernetes.io/audit: restricted
    pod-security.kubernetes.io/audit-version: v1.24
    pod-security.kubernetes.io/warn: restricted
    pod-security.kubernetes.io/warn-version: v1.24
  name: test
  resourceVersion: "104920"
  uid: b76b2aec-0d90-4b70-acaa-f762c78add67
spec:
  finalizers:
  - kubernetes
status:
  phase: Active
```

Reviewer 2: You are right @jianzhangbjz.
This case can NOT be sorted out by the PSA label sync =  true. 

I just tested it out and the OCP matches the pod/image to restricted-v2 so OCP is unable to check that the image requires scale permissions and it fails in one of the impacted case scenarios.

You can ignore my comment and we need to say as workarround just enforce the namespace as privileged.

Reviewer 2: @jiazha@redhat.com you need to apple the label to enforce restricted: 
Example: oc label --overwrite ns <my-catalog-ns> pod-security.kubernetes.io/enforce=restricted.

In your test, the NS is configured to WARNING and AUDIT based on the restricted policy but it is NOT with the enforce.

+
.Example `manifests/<target_namespace>.yaml` file
+
[source,yaml]
----
apiVersion: v1
kind: Namespace
metadata:
...
  labels:
    security.openshift.io/scc.podSecurityLabelSync: "false"
    openshift.io/cluster-monitoring: "true"
    pod-security.kubernetes.io/audit: baseline <1>
    pod-security.kubernetes.io/warn: baseline <2>

>     pod-security.kubernetes.io/enforce: baseline <3>

Reviewer 2: @michaelryanpeter  @anik120 why we are doing these suggestions?
Note that intentionally namespaces prefixed with `openshift-` and a few other ones ( which are created in the installation are considered Openshift system namespace. (see [here](https://github.com/openshift/openshift-docs/pull/48409/files#diff-6b079749a009ff430faa75c3cdece0921ba0e269d1dcd32d6ec0939c7ff99059R11-R23)) Therefore, of these ones is not possible to an enable or disable automatic pod security admission synchronization for most namespaces. 

However, ONLY in 4.12, they will be enforced as restricted. For now, they are privileged. So, I do not understand the relation here with the issue faced. On top of that, I do not think that we should:

a) recommend creating catalog sources or installing Operators on them
b) change the defined enforcement ( they are OCP system namespaces and to aim security we will change the enforcement to restricted on 4.12 )
c) or recommend changing the namespaces directly. We might want to say for users to enable the PSA synchronization. However, I do not think that it has any relation to the issue. 

- We should not recommend users use them to create catalog source
-

> .Example `manifests/<target_namespace>.yaml` file
> +
> [source,yaml]
> ----
> apiVersion: v1
> kind: Namespace
> metadata:
> ...
>   labels:
>     security.openshift.io/scc.podSecurityLabelSync: "false"
>     openshift.io/cluster-monitoring: "true"
>     pod-security.kubernetes.io/audit: baseline <1>
>     pod-security.kubernetes.io/warn: baseline <2>
>     pod-security.kubernetes.io/enforce: baseline <3>
>   name: "<namespace_name>"
> ----
> <1> Set the `audit` mode label to `baseline`.
> <2> Set the `warn` mode label to `baseline`.
> <3> Set the `enforce` mode label to `baseline`.

Reviewer 2: ```suggestion
```
@michaelryanpeter the warnings are only warning then do not cause issues and we do not need to suggest them remove/change it as part of a workaround fix. 

Also, we should recommend using the PSA label syncer instead of changing the enforcement so that we let OCP Auth manage the namespace enforce/audit/warn whatever was it is usually does across the cluster
